Initial Classes

mecha opened the pull request:
The initial abstract classes and traits.

mecha committed:
Added test for `TimestampAwareTrait`

mecha committed:
Added `TimestampAwareTrait`

mecha committed:
Removed signed zeroes in `TimestampAwareTrait` test

Signed zeroes only validate as integers after PHP 5.4.11
Prior to that version, they only validated as floats.

mecha committed:
Updated dependencies

mecha committed:
Added test for `PeriodAwareTrait`

mecha committed:
Added `PeriodAwareTrait`

mecha committed:
Added test for `IntervalAwareTrait`

mecha committed:
Added `IntervalAwareTrait`

mecha committed:
Added PHPUnit config for PHPStorm

mecha committed:
Added test for `AbstractPeriod`

mecha committed:
Added `AbstractPeriod`

mecha committed:
Explicitly set default timezone in test bootstrap

mecha committed:
Updated dependencies

mecha committed:
Added test for `AbstractIteratorRecurringTime`

mecha committed:
Added `AbstractIteratorRecurringTime`

mecha committed:
Added test for `RecurringTimeAwareTrait`

mecha committed:
Added `RecurringTimeAwareTrait`

XedinUnknown commented on src/IntervalAwareTrait.php:
```diff
@@ -0,0 +1,90 @@
+    {
+        if ($interval !== null && !($interval instanceof IntervalInterface)) {
+            throw $this->_createInvalidArgumentException(
+                $this->__('Argument is not a valid interval instance or null.'),
```
Here and everywhere else, something like this is enough:

> Argument is not a valid interval.

mecha replied:
And no period, right? I recall agreeing on no period at the end.

XedinUnknown replied:
That can be a new convention, then: if it's just one sentence, no period.

XedinUnknown commented on src/IntervalAwareTrait.php:
```diff
@@ -0,0 +1,90 @@
+     */
+    protected function _setInterval($interval)
+    {
+        if ($interval !== null && !($interval instanceof IntervalInterface)) {
```
Here and everywhere else, import Time interfaces from `Dhii` namespace.

XedinUnknown commented on src/AbstractPeriod.php:
```diff
@@ -0,0 +1,41 @@
+     *
+     * @since [*next-version*]
+     *
+     * @return int An integer number of seconds. May be negative if the start time is larger than the end time.
```
How can a duration be negative?

mecha replied:
When `start > end`. Though I have no idea why I documented this. While true, mentioning it here is just confusing. Remove?

XedinUnknown commented on src/IntervalAwareTrait.php:
```diff
@@ -0,0 +1,90 @@
+     * Translates a string, and replaces placeholders.
+     *
+     * @since [*next-version*]
+     * @see   sprintf()
```
Here and in other instances of `__()` usage, please remove reference to `sprintf()`.

XedinUnknown commented on src/RecurringTimeAwareTrait.php:
```diff
@@ -0,0 +1,90 @@
+ *
+ * @since [*next-version*]
+ */
+trait RecurringTimeAwareTrait
```
I guess that this is now obsolete.

mecha committed:
Removed `RecurringTimeAwareTrait` and test

mecha committed:
Outsourced dev dependency ...

- Removed `rebelcode/time-interface`
- Installed `dhii/time-interface`

mecha committed:
Fixed FQNs in tests

mecha committed:
Fixed uses of interfaces from `dhii/time-interface`

mecha committed:
Fixed code standards - indentation

mecha committed:
Fixed code standards - unused use statement

mecha committed:
Removed mention of `sprintf()` from ...

... abstract translation methods.

mecha committed:
Shortened invalid argument exception messages

mecha committed:
Updated dependencies

mecha committed:
Merge branch 'develop' into task/initial-classes

mecha committed:
Added test for `SanitizeTimestampCapableTrait`

mecha committed:
Added `SanitizeTimestampCapableTrait`

mecha committed:
Abstracted sanitization in `TimestampAwareTrait`

mecha committed:
Updated dependencies

mecha committed:
Improved test for `TimestampAwareTrait` ...

- Fixed test name
- Added expectation for sanitization
- Removed internal type assertion

mecha committed:
Added test for `StartAwareTrait`

mecha committed:
Added `StartAwareTrait`

mecha committed:
Added missing mock method for `StartAwareTrait` test

mecha committed:
Added test for `EndAwareTrait`

mecha committed:
Added `EndAwareTrait`

mecha committed:
Removed doc for negative duration periods

mecha committed:
Replaced `time()` with `PHP_INT_MAX` in ...

... test for `SanitizeTimestampCapableTrait`

mecha committed:
Fixed PHPUnit config for PHPStorm

mecha committed:
Custom sanitizing, no longer using `filter_var()`

mecha committed:
Fixed random values in tests

mecha committed:
Added test for non-whole number float sanitizing

mecha committed:
Removed unneccessary abstract methods

mecha committed:
Added test for `DurationAwareTrait`

mecha committed:
Added `DurationAwareTrait`

mecha committed:
Improved assertion failure messages

XedinUnknown commented on src/DurationAwareTrait.php:
```diff
@@ -0,0 +1,100 @@
+    protected function _setDuration($duration)
+    {
+        try {
+            $this->duration = $this->_sanitizeTimestamp($duration);
```
Why is `null` not allowed? I don't have a strong opinion about this yet, just asking.

mecha replied:
Good question. Does it make sense to "not have a duration"?

XedinUnknown replied:
Well, everywhere else it is possible to pass `null` - like this you can "unset" the whatever. Also, the initial duration is `null`, so I think not allowing an object to return to that same state can create weird consequences.

XedinUnknown commented on src/DurationAwareTrait.php:
```diff
@@ -0,0 +1,100 @@
+     *
+     * @return int The sanitized timestamp.
+     */
+    abstract protected function _sanitizeTimestamp($timestamp);
```
I think *this* method shouldn't allow `null`: because that indicates an absence of a value, you can't really normalize it to a number. Which is contrary to the setter, which can use `null` to "unset" the value, or simply accept an "empty" value.

XedinUnknown commented on src/SanitizeTimestampCapableTrait.php:
```diff
@@ -0,0 +1,86 @@
+     *
+     * @return int The sanitized timestamp.
+     */
+    protected function _sanitizeTimestamp($timestamp)
```
Why not call [`_normalizeInt()`] here?

[`_normalizeInt()`]: https://github.com/Dhii/normalization-helper-base/blob/task/initial-classes/src/NormalizeIntCapableTrait.php#L30

mecha replied:
I don't think `_normalizeInt()` was available at the time of writing this. I will change it.

mecha replied:
Which sounds better? Having `_sanitizeTimestamp()` call an abstract `_normalizeInt()`? Or replacing all uses of `_sanitizeTimestamp()` with `_normalizeInt()`?

I think the former is a bit useless. It makes the trait require another trait for _all_ of its functionality. What we can do perhaps is make `SanitizeTimestampCapableTrait` `use NormalizeIntCapableTrait`.

Thoughts?

mecha replied:
I'm tending towards replacing all uses with `_normalizeInt()`. While there _may_ be a benefit to having an extra layer of abstraction, these properties (duration, timestamp, ...) are first and foremost integers.

XedinUnknown replied:
I think that to be 100% correct, `_sanitizeTimestamp()` should remain, and should call `_normalizeInt()`. Here, and everywhere else. Another question is that it's additional dev time, but it's still better. The reason for this is that a "timestamp" is not a synonym of "integer", and therefore their definitions may be different. Currently they are the same, but that may change.

mecha replied:
Agree. What about renaming to `_normalizeTimestamp()` instead?

XedinUnknown replied:
Sure.

XedinUnknown commented on src/SanitizeTimestampCapableTrait.php:
```diff
@@ -0,0 +1,86 @@
+     *
+     * @since [*next-version*]
+     *
+     * @param int|string|StringableInterface|null $timestamp The timestamp.
```
Probably `null` shouldn't be sanitizable.

XedinUnknown commented on src/SanitizeTimestampCapableTrait.php:
```diff
@@ -0,0 +1,86 @@
+
+namespace RebelCode\Time;
+
+use Dhii\Util\String\StringableInterface;
```
Looks like too much automation ;)

mecha committed:
Fixed uses in `SanitizeTimestampCapableTrait`

mecha committed:
`DurationAwareTrait` uses `_normalizeTimestamp()`

mecha committed:
Refactored `SanitizeTimestampCapableTrait`

- Now called `NormalizeTimestampCapableTrait`
- No longer accepts null values
- Depends on `_normalizeInt`

mecha committed:
Amended docs

mecha committed:
Fixed code standards for tests

XedinUnknown commented:
Please alias `develop` branch to `0.1.x-dev`.

mecha committed:
Aliased `develop` to `0.1.x-dev`
